CONTRIBUTING.md

```diff
@@ -0,0 +1,227 @@
+not end with whitespaces, this includes code, comments, documentation (`*.md` files), configs and everything else.
+
+Pay attention to existing code and mimic its structure, for instance we keep dependencies sorted order in `Cargo.toml`
+and if you submit changes where new dependency is added out-of-order you're introducing unnecessary entropy and irritate
+maintainers.
+
+If there is an option between being smart with a tricky piece of code or obvious, prefer obvious. Write idiomatic Rust
+as good as you can, this often allows to avoid needing `.unwrap()`.
+
+Struct definition and its implementation should be in the same file, ideally with nothing else in between.
+
+In blockchain code use of `.unwrap()` is strictly forbidden and must never be used (except test code), use `.expect()`
+instead. `.expect()` must convince reviewer that it will never panic or why panicking might be the preferred outcome
+(rarely, but sometimes happens). `.expect()` correctness should be derivable from local context if it is a standalone
+function or invariants of a data structure for internal method. If it can't be guaranteed then `Result<T, E>` must be
+used instead. There are rare exceptions to this such as block numbers that never exceed `u32` in Subspace.
+
+Indexing into slices and similar data structures with `[]` should be avoided and replaced with `.get()` or similar
+methods with explicit handling of cases where element doesn't exist.
+
+Same goes about mathematical operations, especially in runtime logic, in most cases use explicit checked, wrapping or
+saturating math, but don't use saturating math just to make things compile if business logic doesn't expect overflow to
+ever happen, use checked math with `.expect()` in such cases or return an error.
+
+Unsafe code is to be avoided except in special cases. If you can't make code to compile and decide to use `unsafe` to
+bypass compiler, chances are you're doing it wrong and should ask for help to avoid `unsafe`. When you do use `unsafe`
+for valid reasons, just like with `.expect()`, you need to provide a proof convincing reviewer that it is safe to do so
+and can't be exploited with public API from safe Rust.
+
+If code is incomplete or has known issues, make sure to leave a TODO explaining what needs to be done or what isn't
+handled properly. For big things consider creating an issue.
+
+Prefer longer names of variables, variables with 1-3 characters a typically a bad choice. There are notable exceptions
+like using `id` in entity data structure or `i` for a simple iteration over things.
+
+Otherwise, look at existing code and try to do your best.
+
+### Commits
+
+Commits should tell a logical step by step story of the changes and ideally be individually meaningful. Use squashing
+and rebasing during development, reorder commits if it makes more sense that way, there is great tooling out there for
+these kinds of modifications.
```

Comment:
```suggestion
[Commits should tell a logical step by step story of the changes and ideally be individually meaningful](https://github.blog/2022-06-30-write-better-commits-build-better-projects/). Use squashing
and rebasing during development, reorder commits if it makes more sense that way, there is great tooling out there for
these kinds of modifications.
```

This is great ✨ There's a good [blog post](https://github.blog/2022-06-30-write-better-commits-build-better-projects/) from GitHub on this which also dives even further on this(e.g rebasing, bisecting, tooling).

Might be worthwhile linking it here for more context?